File: leetcode/hard/totalAppealOfAString.cpp
using ll = long long;
class Solution {
public:
    long long appealSum(string s) {
        const ll n = s.size(); 
        array<ll, 26> cnt = {};
        ll dp = 0;
        ll res = 0;
        for(int i = 1; i <= n; ++i) {
            auto c = s[i-1];
            dp += (i-cnt[c-'a']);
            cnt[c-'a'] = i;
            res += dp;
        }
        return res;
    }
};

File: leetcode/hard/totalAppealOfAString.py
class Solution:
    def appealSum(self, s: str) -> int:
        n = len(s)
        curr = res = 0
        last = [0] * 26
        for i, c in enumerate(s):
            curr += (i+1) - last[ord(c) - ord('a')]
            res += curr
            last[ord(c) - ord('a')] = i+1
        return res